File: 442_FindAllDuplicatesInAnArray.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDuplicates = function(nums) {
    //Runtime: 108 ms, faster than 64.47% of JavaScript online submissions for Find All Duplicates in an Array.
    //Memory Usage: 43.9 MB, less than 50.00% of JavaScript online submissions for Find All Duplicates in an Array.
    var res=[];
    nums.forEach(function(num){
        var value= Math.abs(num)-1;
        if (nums[value]<0){
            res.push((value+1));
        };
        nums[value]=-nums[value]
    });
    return(res);
};

var inputi =[4,3,2,7,8,2,3,1];

console.log(findDuplicates(inputi));
console.log('-----------the end!----------------');

File: 442_FindAllDuplicatesInAnArray.py
class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        #Runtime: 332 ms, faster than 72.03% of Python online submissions for Find All Duplicates in an Array.
        #Memory Usage: 18.7 MB, less than 66.67% of Python online submissions for Find All Duplicates in an Array.
        # other`s solution: time: o(n) space: o(n)  no extra space
        if not nums: return []
        res=[]
        # use the existing nums to save the status.
        # the value in nums: origal value or - value
        # nums[i] : [1,n]   so nums[abs(value)-1]
        for value in nums:
            if nums[abs(value)-1]<0:
                res.append(abs(value))
            else:
                nums[abs(value)-1]*=-1
        return res


if __name__ == '__main__':
    object = Solution()
    num =  [4,3,2,7,8,2,3,1]
    #num="abab"
    #p="ab"

    print(num)
    print('---Start---')
    r = object.findDuplicates(num)
    print(r)
    print('---End---')
